File: tensor_rt/src/tokenizer.cpp
#include "tokenizer.h"

#include <algorithm>
#include <climits>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <regex>
#include <set>
#include <sstream>

// Build the byte-to-unicode mapping used by CLIP's tokenizer.
//
// The problem: BPE operates on strings, but we need to handle arbitrary bytes
// (including control characters, null bytes, etc.). The solution is to map each
// byte value (0-255) to a unique Unicode character:
//
//   - Printable bytes (33-126, 161-172, 174-255) map to themselves
//     e.g., byte 65 ('A') → "A", byte 97 ('a') → "a"
//
//   - Non-printable bytes (0-32, 127-160, 173) map to codepoints 256+
//     e.g., byte 0 → U+0100 (Ā), byte 32 (space) → U+0120 (Ġ)
//
// This ensures every byte has a visible, unique representation in the
// vocabulary without collisions or ambiguity.
static std::unordered_map<uint8_t, std::string> bytes_to_unicode() {
    std::unordered_map<uint8_t, std::string> result;

    // Collect bytes that map to themselves (printable ranges).
    std::vector<int> bs;
    for (int i = '!'; i <= '~'; i++) bs.push_back(i);      // ASCII printable
    for (int i = 0xA1; i <= 0xAC; i++) bs.push_back(i);    // Latin-1 supplement
    for (int i = 0xAE; i <= 0xFF; i++) bs.push_back(i);    // Latin-1 supplement

    std::set<int> bs_set(bs.begin(), bs.end());

    // Map printable bytes to their Unicode character (encoded as UTF-8).
    for (int b : bs) {
        std::string s;
        if (b < 0x80) {
            s = std::string(1, (char)b);
        } else {
            // 2-byte UTF-8: 110xxxxx 10xxxxxx
            s += (char)(0xC0 | (b >> 6));
            s += (char)(0x80 | (b & 0x3F));
        }
        result[(uint8_t)b] = s;
    }

    // Map remaining bytes to codepoints starting at 256.
    int n = 0;
    for (int b = 0; b < 256; b++) {
        if (bs_set.count(b) == 0) {
            int cp = 256 + n;
            std::string s;
            if (cp < 0x80) {
                s = std::string(1, (char)cp);
            } else if (cp < 0x800) {
                s += (char)(0xC0 | (cp >> 6));
                s += (char)(0x80 | (cp & 0x3F));
            }
            result[(uint8_t)b] = s;
            n++;
        }
    }

    return result;
}

bool ClipTokenizer::load(const std::string& vocab_path) {
    byte_encoder_ = bytes_to_unicode();

    std::ifstream file(vocab_path);
    if (!file.is_open()) {
        std::cerr << "Failed to open vocab: " << vocab_path << std::endl;
        return false;
    }

    std::string line;

    // First line is a header comment, skip it.
    std::getline(file, line);

    // Read BPE merge rules. Each line is "token1 token2", meaning those two
    // tokens should be merged (in priority order — earlier = merge first).
    // CLIP uses exactly 48,894 merges (not the full 49,152 in the file).
    const int max_merges = 48894;
    int rank = 0;
    while (std::getline(file, line) && rank < max_merges) {
        if (line.empty()) continue;

        auto space = line.find(' ');
        if (space == std::string::npos) continue;

        std::string first = line.substr(0, space);
        std::string second = line.substr(space + 1);

        std::string key = first + " " + second;
        bpe_ranks_[key] = rank++;
    }
    file.close();

    // Build the vocabulary (token string → integer ID).
    //
    // The vocab is constructed in a specific order that determines token IDs:
    //   IDs 0-255:     single-byte tokens (one per byte value)
    //   IDs 256-511:   single-byte + "</w>" tokens (end-of-word variants)
    //   IDs 512-49405: merged tokens (one per merge rule, in merge order)
    //   ID 49406:      <|startoftext|>
    //   ID 49407:      <|endoftext|>
    std::vector<std::string> vocab;

    // Single-byte tokens (IDs 0-255).
    for (int b = 0; b < 256; b++) {
        auto it = byte_encoder_.find((uint8_t)b);
        if (it != byte_encoder_.end()) {
            vocab.push_back(it->second);
        }
    }

    // Byte + </w> tokens (IDs 256-511).
    for (int b = 0; b < 256; b++) {
        auto it = byte_encoder_.find((uint8_t)b);
        if (it != byte_encoder_.end()) {
            vocab.push_back(it->second + "</w>");
        }
    }

    // Merged tokens (IDs 512+). Each merge rule creates a new token by
    // concatenating the two tokens it merges.
    std::ifstream file2(vocab_path);
    std::getline(file2, line);  // skip header

    int merge_count = 0;
    while (std::getline(file2, line) && merge_count < max_merges) {
        if (line.empty()) continue;
        auto space = line.find(' ');
        if (space == std::string::npos) continue;
        merge_count++;

        std::string merged = line.substr(0, space) + line.substr(space + 1);
        vocab.push_back(merged);
    }
    file2.close();

    // Special tokens.
    vocab.push_back("<|startoftext|>");
    vocab.push_back("<|endoftext|>");

    for (int i = 0; i < (int)vocab.size(); i++) {
        encoder_[vocab[i]] = i;
    }

    sot_token_ = encoder_["<|startoftext|>"];
    eot_token_ = encoder_["<|endoftext|>"];

    std::cout << "Loaded CLIP tokenizer: " << encoder_.size() << " tokens, "
              << bpe_ranks_.size() << " merges" << std::endl;
    return true;
}

std::vector<std::string> ClipTokenizer::byte_encode(
    const std::string& text) const {
    std::vector<std::string> result;
    for (unsigned char c : text) {
        auto it = byte_encoder_.find(c);
        if (it != byte_encoder_.end()) {
            result.push_back(it->second);
        }
    }
    return result;
}

// Apply BPE merges to a single word.
//
// Starting from individual characters (byte-encoded), repeatedly find and
// merge the highest-priority pair until no more merges apply.
//
// Example for "space":
//   Initial:  ["s", "p", "a", "c", "e</w>"]
//   Merge 1:  ["sp", "a", "c", "e</w>"]       (rank of "s p" is lowest)
//   Merge 2:  ["sp", "a", "ce</w>"]            (rank of "c e</w>")
//   Merge 3:  ["sp", "ace</w>"]                (rank of "a ce</w>")
//   Merge 4:  ["space</w>"]                    (rank of "sp ace</w>")
//   Done: no more pairs have merge rules.
std::vector<std::string> ClipTokenizer::bpe(const std::string& token) const {
    // Split into individual UTF-8 characters.
    std::vector<std::string> word;
    for (size_t i = 0; i < token.size();) {
        // Determine UTF-8 character length from first byte.
        int len = 1;
        unsigned char c = token[i];
        if ((c & 0xE0) == 0xC0) len = 2;
        else if ((c & 0xF0) == 0xE0) len = 3;
        else if ((c & 0xF8) == 0xF0) len = 4;

        word.push_back(token.substr(i, len));
        i += len;
    }

    if (word.empty()) return word;

    // Mark the last character with </w> (end-of-word).
    // This distinguishes "space" (end of word) from "space" (prefix of
    // "spaceship"), which get different BPE tokens.
    word.back() += "</w>";

    // Greedily merge pairs until no more merges apply.
    while (word.size() > 1) {
        // Find the pair with the lowest merge rank (highest priority).
        int best_rank = INT_MAX;
        int best_idx = -1;

        for (size_t i = 0; i + 1 < word.size(); i++) {
            std::string pair_key = word[i] + " " + word[i + 1];
            auto it = bpe_ranks_.find(pair_key);
            if (it != bpe_ranks_.end() && it->second < best_rank) {
                best_rank = it->second;
                best_idx = i;
            }
        }

        if (best_idx < 0) break;  // No mergeable pairs remain.

        // Merge the best pair into a single token.
        std::string merged = word[best_idx] + word[best_idx + 1];
        std::vector<std::string> new_word;
        for (size_t i = 0; i < word.size(); i++) {
            if ((int)i == best_idx) {
                new_word.push_back(merged);
                i++;  // skip the second token of the merged pair
            } else {
                new_word.push_back(word[i]);
            }
        }
        word = std::move(new_word);
    }

    return word;
}

std::vector<int32_t> ClipTokenizer::encode(const std::string& text,
                                            int max_length) const {
    std::vector<int32_t> tokens;

    // Start with <|startoftext|> token (always the first token).
    tokens.push_back(sot_token_);

    // Lowercase the input (CLIP was trained on lowercased text).
    std::string clean;
    for (char c : text) {
        clean += std::tolower((unsigned char)c);
    }

    // Split on whitespace, then byte-encode and BPE each word.
    //
    // Note: the real CLIP tokenizer uses a regex to split on word boundaries
    // and punctuation. This simplified version splits on spaces only, which
    // works fine for typical prompts like "Taj Mahal in space".
    std::string word;
    for (size_t i = 0; i < clean.size(); i++) {
        char c = clean[i];
        if (c == ' ' || c == '\t' || c == '\n') {
            if (!word.empty()) {
                // Convert word bytes to Unicode representations, then run BPE.
                std::string encoded;
                for (unsigned char ch : word) {
                    auto it = byte_encoder_.find(ch);
                    if (it != byte_encoder_.end()) {
                        encoded += it->second;
                    }
                }
                auto bpe_tokens = bpe(encoded);
                for (const auto& t : bpe_tokens) {
                    auto it = encoder_.find(t);
                    if (it != encoder_.end()) {
                        tokens.push_back(it->second);
                    }
                }
                word.clear();
            }
        } else {
            word += c;
        }
    }

    // Process the last word (no trailing space).
    if (!word.empty()) {
        std::string encoded;
        for (unsigned char ch : word) {
            auto it = byte_encoder_.find(ch);
            if (it != byte_encoder_.end()) {
                encoded += it->second;
            }
        }
        auto bpe_tokens = bpe(encoded);
        for (const auto& t : bpe_tokens) {
            auto it = encoder_.find(t);
            if (it != encoder_.end()) {
                tokens.push_back(it->second);
            }
        }
    }

    // End with <|endoftext|> token.
    tokens.push_back(eot_token_);

    // CLIP always expects exactly 77 tokens. Truncate if too long,
    // pad with <|endoftext|> if too short.
    if ((int)tokens.size() > max_length) {
        tokens.resize(max_length);
        tokens.back() = eot_token_;
    }
    while ((int)tokens.size() < max_length) {
        tokens.push_back(eot_token_);
    }

    return tokens;
}

File: zig_tensorrt/src/text_to_image/tokenizer.zig
// CLIP BPE (Byte-Pair Encoding) tokenizer — pure Zig.
//
// ┌─────────────────────────────────────────────────────────────────────────┐
// │  WHAT THIS DOES                                                         │
// │                                                                         │
// │  Neural networks only understand numbers. Before the text encoder can   │
// │  process a prompt like "Taj Mahal in space", it must be converted to    │
// │  a sequence of integer token IDs:                                       │
// │                                                                         │
// │    "Taj Mahal in space"                                                 │
// │        ↓ lowercase                                                      │
// │    "taj mahal in space"                                                 │
// │        ↓ split on whitespace                                            │
// │    ["taj", "mahal", "in", "space"]                                      │
// │        ↓ byte-encode each word → BPE merge → vocab lookup               │
// │    [24805, 22301, 530, 2138]                                            │
// │        ↓ wrap with start/end tokens, pad to 77                          │
// │    [49406, 24805, 22301, 530, 2138, 49407, 49407, ...]                  │
// │                                                                         │
// │  HOW BPE WORKS                                                          │
// │                                                                         │
// │  BPE is a compression algorithm. Starting from individual characters,   │
// │  it repeatedly merges the most common adjacent pair:                    │
// │                                                                         │
// │    "space" → ["s", "p", "a", "c", "e</w>"]                             │
// │            → ["sp", "a", "c", "e</w>"]       merge "s"+"p"             │
// │            → ["sp", "a", "ce</w>"]            merge "c"+"e</w>"         │
// │            → ["sp", "ace</w>"]                merge "a"+"ce</w>"        │
// │            → ["space</w>"]                    merge "sp"+"ace</w>"      │
// │                                                                         │
// │  The </w> suffix marks end-of-word, so "space" at the end of a word    │
// │  gets a different token than "space" as a prefix of "spaceship".        │
// │                                                                         │
// │  WHY BYTE-LEVEL ENCODING?                                               │
// │                                                                         │
// │  To handle ANY input (Unicode, control chars, etc.), each byte (0-255)  │
// │  is mapped to a unique visible Unicode character. Printable ASCII maps  │
// │  to itself (A→A), while non-printable bytes get mapped to codepoints    │
// │  starting at 256 (space→Ġ, null→Ā). This avoids "unknown token"       │
// │  issues entirely.                                                       │
// │                                                                         │
// │  VOCABULARY STRUCTURE (49,408 tokens total):                            │
// │    IDs 0-255:     single-byte tokens                                    │
// │    IDs 256-511:   single-byte + "</w>" tokens                           │
// │    IDs 512-49405: merged tokens (one per merge rule)                    │
// │    ID 49406:      <|startoftext|>                                       │
// │    ID 49407:      <|endoftext|>                                         │
// └─────────────────────────────────────────────────────────────────────────┘

const std = @import("std");

const max_merges = 48894;
const max_token_len = 77;
const sot_token: i32 = 49406;
const eot_token: i32 = 49407;

pub const ClipTokenizer = struct {
    allocator: std.mem.Allocator,
    arena: std.heap.ArenaAllocator,

    /// token string → integer ID (e.g. "space</w>" → 2330)
    encoder: std.StringHashMap(i32),
    /// "first second" → merge priority rank (lower = merge first)
    bpe_ranks: std.StringHashMap(i32),
    /// byte value (0-255) → Unicode string representation
    byte_encoder: [256][]const u8,

    pub fn init(allocator: std.mem.Allocator) ClipTokenizer {
        var self = ClipTokenizer{
            .allocator = allocator,
            .arena = std.heap.ArenaAllocator.init(allocator),
            .encoder = std.StringHashMap(i32).init(allocator),
            .bpe_ranks = std.StringHashMap(i32).init(allocator),
            .byte_encoder = undefined,
        };
        // Initialize byte_encoder to empty slices
        for (&self.byte_encoder) |*entry| {
            entry.* = "";
        }
        return self;
    }

    pub fn deinit(self: *ClipTokenizer) void {
        self.encoder.deinit();
        self.bpe_ranks.deinit();
        self.arena.deinit();
    }

    /// Load the BPE vocabulary file (bpe_simple_vocab_16e6.txt).
    pub fn load(self: *ClipTokenizer, vocab_path: []const u8) !void {
        self.buildByteEncoder();

        // Read the entire file into memory (~3.2MB)
        const file = try std.fs.cwd().openFile(vocab_path, .{});
        defer file.close();
        const file_size = (try file.stat()).size;
        const contents = try self.arena.allocator().alloc(u8, file_size);
        var total_read: usize = 0;
        while (total_read < file_size) {
            const n = try file.read(contents[total_read..]);
            if (n == 0) break;
            total_read += n;
        }
        const data = contents[0..total_read];

        // Split into lines
        var lines = std.mem.splitScalar(u8, data, '\n');

        // Skip header line
        _ = lines.next();

        // First pass: read BPE merge rules
        var rank: i32 = 0;
        var lines_copy = lines; // save position for second pass
        while (rank < max_merges) {
            const l = lines.next() orelse break;
            if (l.len == 0) continue;
            const space_idx = std.mem.indexOfScalar(u8, l, ' ') orelse continue;
            const key = try std.fmt.allocPrint(self.arena.allocator(), "{s} {s}", .{ l[0..space_idx], l[space_idx + 1 ..] });
            try self.bpe_ranks.put(key, rank);
            rank += 1;
        }

        // Build vocabulary in the specific order that determines token IDs.
        // Single-byte tokens (IDs 0-255)
        for (0..256) |b| {
            const s = self.byte_encoder[b];
            if (s.len > 0) {
                try self.encoder.put(s, @intCast(b));
            }
        }

        // Byte + </w> tokens (IDs 256-511)
        for (0..256) |b| {
            const s = self.byte_encoder[b];
            if (s.len > 0) {
                const sw = try std.fmt.allocPrint(self.arena.allocator(), "{s}</w>", .{s});
                try self.encoder.put(sw, @as(i32, @intCast(b)) + 256);
            }
        }

        // Second pass: merged tokens (IDs 512+)
        var merge_id: i32 = 512;
        var mc: i32 = 0;
        while (mc < max_merges) {
            const l = lines_copy.next() orelse break;
            if (l.len == 0) continue;
            const space_idx = std.mem.indexOfScalar(u8, l, ' ') orelse continue;
            mc += 1;
            const merged = try std.fmt.allocPrint(self.arena.allocator(), "{s}{s}", .{ l[0..space_idx], l[space_idx + 1 ..] });
            try self.encoder.put(merged, merge_id);
            merge_id += 1;
        }

        // Special tokens
        const sot_str = try self.arena.allocator().dupe(u8, "<|startoftext|>");
        const eot_str = try self.arena.allocator().dupe(u8, "<|endoftext|>");
        try self.encoder.put(sot_str, sot_token);
        try self.encoder.put(eot_str, eot_token);

        std.debug.print("Loaded CLIP tokenizer: {d} tokens, {d} merges\n", .{ self.encoder.count(), self.bpe_ranks.count() });
    }

    /// Tokenize text into a fixed-length 77-token sequence:
    /// [<|startoftext|>, ...tokens..., <|endoftext|>, pad, pad, ...]
    pub fn encode(self: *ClipTokenizer, text: []const u8, buf: []i32) void {
        @memset(buf, eot_token);
        buf[0] = sot_token;
        var count: usize = 1;

        // Lowercase the input (CLIP was trained on lowercased text)
        var lower: [512]u8 = undefined;
        const text_len = @min(text.len, lower.len);
        for (0..text_len) |i| {
            lower[i] = std.ascii.toLower(text[i]);
        }
        const clean = lower[0..text_len];

        // Split on whitespace and process each word
        var iter = std.mem.tokenizeAny(u8, clean, " \t\n");
        while (iter.next()) |word| {
            if (count >= max_token_len - 1) break; // leave room for EOT

            // Byte-encode the word: convert each byte to its Unicode representation
            var encoded_buf: [2048]u8 = undefined;
            var encoded_len: usize = 0;
            for (word) |ch| {
                const s = self.byte_encoder[ch];
                @memcpy(encoded_buf[encoded_len..][0..s.len], s);
                encoded_len += s.len;
            }
            const encoded = encoded_buf[0..encoded_len];

            // Apply BPE merges, then look up each resulting token in the vocabulary
            var bpe_result: [64][]const u8 = undefined;
            const bpe_count = self.bpe(encoded, &bpe_result);

            for (bpe_result[0..bpe_count]) |token| {
                if (count >= max_token_len - 1) break;
                if (self.encoder.get(token)) |id| {
                    buf[count] = id;
                    count += 1;
                }
            }
        }

        // Ensure EOT is placed after the last real token
        if (count < max_token_len) {
            buf[count] = eot_token;
        }
    }

    /// Build the byte-to-unicode mapping (same as OpenAI's bytes_to_unicode).
    fn buildByteEncoder(self: *ClipTokenizer) void {
        const alloc = self.arena.allocator();

        // Printable ASCII and Latin-1 supplement map to themselves
        var is_printable = [_]bool{false} ** 256;
        for ('!'..('~' + 1)) |b| is_printable[b] = true;
        for (0xA1..0xAD) |b| is_printable[b] = true;
        for (0xAE..0x100) |b| is_printable[b] = true;

        for (0..256) |b| {
            if (is_printable[b]) {
                // Map to the Unicode codepoint for this byte value
                if (b < 0x80) {
                    self.byte_encoder[b] = alloc.dupe(u8, &[_]u8{@intCast(b)}) catch "";
                } else {
                    // 2-byte UTF-8 encoding
                    const bytes = [_]u8{
                        @intCast(0xC0 | (b >> 6)),
                        @intCast(0x80 | (b & 0x3F)),
                    };
                    self.byte_encoder[b] = alloc.dupe(u8, &bytes) catch "";
                }
            }
        }

        // Non-printable bytes get mapped to codepoints starting at 256
        var n: usize = 0;
        for (0..256) |b| {
            if (!is_printable[b]) {
                const cp = 256 + n;
                if (cp < 0x80) {
                    self.byte_encoder[b] = alloc.dupe(u8, &[_]u8{@intCast(cp)}) catch "";
                } else if (cp < 0x800) {
                    const bytes = [_]u8{
                        @intCast(0xC0 | (cp >> 6)),
                        @intCast(0x80 | (cp & 0x3F)),
                    };
                    self.byte_encoder[b] = alloc.dupe(u8, &bytes) catch "";
                }
                n += 1;
            }
        }
    }

    /// Apply BPE merges to a single word.
    /// Starts from individual UTF-8 characters, adds </w> to last,
    /// then greedily merges the highest-priority pair until done.
    fn bpe(self: *ClipTokenizer, token: []const u8, result: *[64][]const u8) usize {
        // Split into individual UTF-8 characters
        var word: [128][]const u8 = undefined;
        var word_len: usize = 0;
        var i: usize = 0;
        while (i < token.len and word_len < 127) {
            // Determine UTF-8 character length from the first byte
            var char_len: usize = 1;
            const c = token[i];
            if (c & 0xE0 == 0xC0) char_len = 2 else if (c & 0xF0 == 0xE0) char_len = 3 else if (c & 0xF8 == 0xF0) char_len = 4;
            if (i + char_len > token.len) break;
            word[word_len] = token[i .. i + char_len];
            word_len += 1;
            i += char_len;
        }
        if (word_len == 0) return 0;

        // Append </w> to the last character (marks end of word)
        var last_buf: [32]u8 = undefined;
        const last = word[word_len - 1];
        @memcpy(last_buf[0..last.len], last);
        @memcpy(last_buf[last.len..][0..4], "</w>");
        // We need a stable pointer — use the arena
        const last_with_w = self.arena.allocator().dupe(u8, last_buf[0 .. last.len + 4]) catch {
            result[0] = token;
            return 1;
        };
        word[word_len - 1] = last_with_w;

        // Greedily merge pairs until no more merges apply
        var pair_buf: [256]u8 = undefined;
        while (word_len > 1) {
            // Find the pair with the lowest merge rank
            var best_rank: i32 = std.math.maxInt(i32);
            var best_idx: usize = word_len; // sentinel = no match

            for (0..word_len - 1) |j| {
                const a = word[j];
                const b = word[j + 1];
                if (a.len + 1 + b.len > pair_buf.len) continue;
                @memcpy(pair_buf[0..a.len], a);
                pair_buf[a.len] = ' ';
                @memcpy(pair_buf[a.len + 1 ..][0..b.len], b);
                const pair_key = pair_buf[0 .. a.len + 1 + b.len];

                if (self.bpe_ranks.get(pair_key)) |r| {
                    if (r < best_rank) {
                        best_rank = r;
                        best_idx = j;
                    }
                }
            }

            if (best_idx >= word_len) break; // No more merges

            // Merge the pair at best_idx
            const a = word[best_idx];
            const b_tok = word[best_idx + 1];
            const merged = self.arena.allocator().alloc(u8, a.len + b_tok.len) catch break;
            @memcpy(merged[0..a.len], a);
            @memcpy(merged[a.len..][0..b_tok.len], b_tok);

            // Rebuild word array with the merged token replacing the pair
            var new_word: [128][]const u8 = undefined;
            var nw: usize = 0;
            var k: usize = 0;
            while (k < word_len) {
                if (k == best_idx) {
                    new_word[nw] = merged;
                    nw += 1;
                    k += 2; // skip both parts of the merged pair
                } else {
                    new_word[nw] = word[k];
                    nw += 1;
                    k += 1;
                }
            }
            @memcpy(word[0..nw], new_word[0..nw]);
            word_len = nw;
        }

        @memcpy(result[0..word_len], word[0..word_len]);
        return word_len;
    }
};
